### PR TITLE
Add concat ops

### PR DESCRIPTION
**Pull request by xytintel**
e.g. cat, concat, concatenate

### CODE REVIEW

**Review by EikanWang — COMMENTED** [review]
Overall, LGTM. Please refine the comments.

File: src/aten/sycl/CatKernel.cpp [hunk 1]
```diff
@@ -0,0 +1,501 @@
...
+  // many threads from needlessly load meta data if their sizes is small.
+
+  return std::make_tuple(2LL * sum_ss, (long long)nTensors);
+}
+
+template <typename T>
+inline std::tuple<unsigned int, unsigned int, unsigned int> get_cat_range_contig(
+    unsigned int max_elements_per_tensor,
+    ptrdiff_t nTensors) {
+  constexpr unsigned int wi_per_group = 256;
```
[the remark this answers is not in the document]
Comment (xytintel) [reply src/aten/sycl/CatKernel.cpp:204]: yes

File: src/aten/sycl/CatKernel.cpp [hunk 2]
```diff
@@ -0,0 +1,501 @@
...
+  constexpr unsigned int min_aligned_vec_per_wi = 1;
+  constexpr unsigned int max_tb_per_ss = 32;
```
[the remark this answers is not in the document]
Comment (xytintel) [reply src/aten/sycl/CatKernel.cpp:206]: tb: thread block, ss: sub-slice

File: src/aten/Cat.cpp [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+}
+
+} // namespace impl
+
+Tensor& XPUNativeFunctions::cat_out(
+    const ITensorListRef& tensors,
+    int64_t dim,
+    Tensor& result) {
+  auto materialized = tensors.materialize();
+  dim = at::legacy_cat_wrap_dim(dim, materialized);
```
[the remark this answers is not in the document]
Comment (xytintel) [reply src/aten/Cat.cpp:40]: No we should not

File: src/aten/TensorShape.cpp [hunk 4]
```diff
@@ -44,4 +52,183 @@ Tensor XPUNativeFunctions::unfold(
   return at::native::unfold(self, dimension, size, step);
 }
 
+inline c10::MemoryFormat cat_compute_output_memory_format(
```
Comment (gujinghui) [inline src/aten/TensorShape.cpp:55]: should we reuse these utilities from PyTorch, instead of copy out?

Comment (xytintel) [reply src/aten/TensorShape.cpp:55]: The code is inside the cu file

Comment (EikanWang) [reply src/aten/TensorShape.cpp:55]: @xytintel , I think we can submit another PR to the stock PyTorch by exposing the utility function. But it should not block this PR.

**Review by EikanWang — APPROVED** [review]

File: src/aten/TensorShape.cpp (hunk 4 above)
Comment (fengyuan14) [reply src/aten/TensorShape.cpp:55]: Will submit a PR to stock PyTorch.

**Pull request merged**